### PR TITLE
catMultiFa

### PR DESCRIPTION
**Pull request by rimangan**
catMultiFa is a new cmd that works like the Unix cat cmd for multiple alignment fasta (multiFa) format files. For multiple multifa files with the same number and order of species, the individual sequences are concatenated to retain multiFa format and produce composite sequences. Useful for downstream population analysis for multiFa alignments from multiple regions. 
Accepts an arbitrary number of multiFa files, supporting syntax like:
"~/go/bin/catMultiFa *.fa"
Default output to stdout, but the cmd accepts an -o flag to redirect output to a file.
The user can also provide a list of filenames as a newline-delimited text file. This is useful when catting thousands of files to avoid the command line character limit.
Passes tests.

### CODE REVIEW

File: cmd/catMultiFa/catMultiFa.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+			"Usage:\n" +
+			"catMultiFa multi.fa ...\n" +
+			"options:\n")
+	flag.PrintDefaults()
+}
+
+func main() {
+	var minNumArgs int = 2
+	var o *string = flag.String("o", "stdout", "Redirect the output to a user-given filename.")
+	var list *string = flag.String("list", "", "")
```
Comment (dasnellings) [inline cmd/catMultiFa/catMultiFa.go:55]: You should point out that this needs to be a line delimited text file. I see it from the code, but I don't think it is obvious if you are just looking at the help message.

File: cmd/catMultiFa/catMultiFa.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	if len(flag.Args()) < minNumArgs {
+		flag.Usage()
+		log.Fatalf("Error: expecting at least %d arguments, but got %d\n", minNumArgs, len(flag.Args()))
+	}
+	var fileList []string
+	var err error
+	if *list == "" {
+		fileList = make([]string, len(flag.Args()))
```
Comment (dasnellings) [inline cmd/catMultiFa/catMultiFa.go:73]: `fileList = flag.Args()`?

**Review by dasnellings — DISMISSED** [review]
Looks good. Though I don't really understand the use case. Making composite sequences is a bit weird.

File: cmd/catMultiFa/catMultiFa.go (hunk 2 above)
Comment (rimangan) [reply cmd/catMultiFa/catMultiFa.go:73]: lol way simpler

**Comment by rimangan** [comment]
> Looks good. Though I don't really understand the use case. Making composite sequences is a bit weird.

This is indeed a bit niche. The current use case is for PCA on HAQERs, where we want to cluster sequences from multiple discontinuous aligned regions at once. The idea here is to use the multiFaExtract -bed to pull all alignments of HAQERs from the genome-wide alignment, and follow up with this to make one continuous sequence that can then feed into PCA. I could imagine this being used to analyze multiple discontinuous sequences together for other types of population analysis, Dunn or Tajima for example

File: cmd/catMultiFa/catMultiFa.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+			"Usage:\n" +
+			"catMultiFa multi.fa ...\n" +
+			"OR\n" +
+			"catMultiFa -list file.list" +
+			"options:\n")
+	flag.PrintDefaults()
+}
+
+func main() {
+	var minNumArgs int = 2
```
Comment (cshoben) [inline cmd/catMultiFa/catMultiFa.go:55]: Can you talk me through the logic behind the number of args? Like between here and how we reset it to zero if the user gives a list. Is it set to 2 here so the math at line 68 always works out?

**Review by cshoben — APPROVED** [review]
Awesome! And also I appreciate this bite size PR. A logic questions included with this approval.

File: cmd/catMultiFa/catMultiFa.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	var fileList []string
+	var err error
+	if *list == "" {
+		fileList = flag.Args()
+	} else {
+		file := fileio.EasyOpen(*list)
+		fileList = make([]string, 0)
+		var line string
+		var doneReading bool
+		for line, doneReading = fileio.EasyNextRealLine(file); !doneReading; line, doneReading = fileio.EasyNextRealLine(file) {
```
Comment (craiglowe) [inline cmd/catMultiFa/catMultiFa.go:81]: I think there is code that does this already (take a filename and read it into a slice where each line of the file is an element in the slice.  It is fileio.Read

**Review by craiglowe — COMMENTED** [review]
Looks great.  I had just a few very small comments

File: cmd/catMultiFa/catMultiFa.go (hunk 3 above)
Comment (rimangan) [reply cmd/catMultiFa/catMultiFa.go:55]: I'm clarifying the logic in the next commit, but the idea here is if you send files as arguments you need at least 2 in order to cat them together. If you use a file list instead, you do not need any arguments (in fact, it's forbidden in the next commit)

File: cmd/catMultiFa/catMultiFa.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	var o *string = flag.String("o", "stdout", "Redirect the output to a user-given filename.")
+	var list *string = flag.String("list", "", "User may supply a line-delimited list of files to concatenate instead of providing them as arguments.")
+	var lineLength *int = flag.Int("lineLength", 50, "Specify the line length in the output file.")
+
+	flag.Usage = usage
+	log.SetFlags(log.Ldate | log.Ltime | log.Lshortfile)
+	flag.Parse()
+
+	if *list != "" {
+		minNumArgs = 0
```
[the remark this answers is not in the document]
Comment (rimangan) [reply cmd/catMultiFa/catMultiFa.go:65]: The user cannot do both, so expectedNumArgs is 0 in this case. I'll rework in the next commit

File: cmd/catMultiFa/catMultiFa.go (hunk 4 above)
Comment (rimangan) [reply cmd/catMultiFa/catMultiFa.go:81]: Great! Changing in next commit

**Review by craiglowe — APPROVED** [review]
Great work!

**Review by dasnellings — APPROVED** [review]

**Review by cshoben — APPROVED** [review]

**Pull request merged**